Add pending refund notes

htdat opened the pull request:
Fixes #1195

#### Changes proposed in this Pull Request

Changes based on this spike https://github.com/Automattic/woocommerce-payments/issues/1195#issuecomment-2677422321: 

- Modify `add_note_and_metadata_for_refund`: change its name and add one param `$is_pending` to `add_note_and_metadata_for_created_refund` so that it can handle both `pending` and `succeeded` note. 
- No need to change on the timeline refund, i.e. [WC_REST_Payments_Refunds_Controller::process_refund](https://github.com/Automattic/woocommerce-payments/blob/fd6b5401620300ad04636a53016dc01a7dc43fb7/includes/admin/class-wc-rest-payments-refunds-controller.php#L46), because it's eventually tunneld to gateway `process_refund` with `wc_create_refund`.
- Update webhook handling so that it can handle both `pending` and `succeeded`. The most notable change is on `process_webhook_refund_updated` where I moved the handled the failed refund into `Order_Service`. 

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

* Use https://docs.stripe.com/testing?testing-method=card-numbers#refunds to test for the relevant flows. (If testing with a local dev server, make sure to enable webhook listening). 
* Trigger pending -> succeeded refund status in all flows: from Stripe dash, WP Admin order admin page, and Transaction Timeline. 
* Trigger succeeded -> failed refund status in the same flows.
* Expect: all respective pending, succeeded, failed notes are added into the order.

-------------------


- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

> [!IMPORTANT]
> Add documents for this new link https://woocommerce.com/document/woopayments/managing-money/#pending-refunds 
> _Update: asked here p1743136317673289-slack-C03EE4CPPK5_

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Fix for gateway + rest api flows

htdat committed:
Fix for webhook

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10600` or branch name `fix/1195-pending-refund` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/1195-pending-refund) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: d27e3007a05bfd1ef62e7f551681f64e74fc4dde
- Build time: 2025-03-26 04:27:10 UTC

_Note: the build is updated when a new commit is pushed to this PR._

htdat committed:
Refactor to use $is_pending

htdat committed:
Refactor to handle both failed and succeeded status for `process_webhook_refund_updated`

htdat committed:
Refactor handle_insufficient_balance_for_refund to use has_webhook_property

htdat committed:
Introduce Refund_Status

- Remove the un-used code in a prev commit

htdat committed:
Merge branch 'develop' into fix/1195-pending-refund

htdat commented:
FYI ready for the initial review. The unit tests are still in progress, but I would love to have initial feedback meanwhile.

htdat committed:
Include Refund Status file + fix test_process_refund_should_work_without_payment_method_id_meta

- changelog

mgascam commented:
@htdat I noticed the following error while testing. It saw it when I tried to visit the order details page, after a checkout:
```
Fatal error: Uncaught Error: Class "WCPay\Constants\Refund_Status" not found in /var/www/html/wp-content/plugins/woocommerce-payments/includes/class-wc-payment-gateway-wcpay.php:2327 Stack trace: #0 /var/www/html/wp-content/plugins/woocommerce-payments/includes/admin/class-wc-payments-admin.php(833): WC_Payment_Gateway_WCPay->has_refund_failed(Object(Automattic\WooCommerce\Admin\Overrides\Order)) #1 /var/www/html/wp-includes/class-wp-hook.php(324): WC_Payments_Admin->enqueue_payments_scripts('woocommerce_pag...') #2 /var/www/html/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array) #3 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action(Array) #4 /var/www/html/wp-admin/admin-header.php(118): do_action('admin_enqueue_s...', 'woocommerce_pag...') #5 /var/www/html/wp-admin/admin.php(239): require_once('/var/www/html/w...') #6 {main} thrown in /var/www/html/wp-content/plugins/woocommerce-payments/includes/class-wc-payment-gateway-wcpay.php on line 2327
``` 
I'm a bit confused, since the class should be available, given the call to `use WCPay\Constants\Refund_Status;`. Also we are already using other constants in the same file 🤔 Sharing in case it rings a bell.

mgascam reviewed (changes requested):
I think we are missing to include the new constant file for the refund status:

```
diff --git a/includes/class-wc-payments.php b/includes/class-wc-payments.php
index 41759213d..321b1f1b4 100644
--- a/includes/class-wc-payments.php
+++ b/includes/class-wc-payments.php
@@ -473,6 +473,7 @@ class WC_Payments {
                include_once __DIR__ . '/exceptions/class-invalid-address-exception.php';
                include_once __DIR__ . '/constants/class-base-constant.php';
                include_once __DIR__ . '/constants/class-country-code.php';
+               include_once __DIR__ . '/constants/class-refund-status.php';
                include_once __DIR__ . '/constants/class-country-test-cards.php';
                include_once __DIR__ . '/constants/class-currency-code.php';
                include_once __DIR__ . '/constants/class-fraud-meta-box-type.php';
```

htdat committed:
Fix + add tests WC_Payments_Webhook_Processing_Service_Test

htdat committed:
Merge branch 'refs/heads/develop' into fix/1195-pending-refund

htdat commented:
@mgascam - yeah, I missed adding the file into the include list. The PR is ready to review. I've added tests covering for the webhook processing change; but I will still need to add tests for changes in Order Service. 

In the meantime, you can review the code too :D 

**_Edit_: with the most recent commit, all tests have been created and updated for changes.**

htdat committed:
Add tests for WC_Payments_Order_Service::handle_failed_refund

htdat committed:
Add tests for changes in WC_Payments_Order_Service::add_note_and_metadata_for_created_refund

mgascam commented:
Hi @htdat I wanted to share one scenario I found while testing the PR. I did the checkout with the Asynchronous success card. The thing is my test account did not have balance to process the refund, so it was pending and I canceled it in the Stripe dashboard. This is the payment for your reference: https://dashboard.stripe.com/test/connect/accounts/acct_1PxRhsR4vt8oOW8F/payments/pi_3R6dCwR4vt8oOW8F1rjgCeW9. If you have a look, you can see that the last triggered event was `refund.failed`. The status of this event is "canceled", but I don't think we are handling it. As a result, things are out of sync and the last order note in my test store is:
<img width="282" alt="Screenshot 2025-03-25 at 20 46 40" src="https://github.com/user-attachments/assets/69f12d4b-13c8-42ee-8136-4fb79a9ae94c" />

Sorry It's a bit late for me to debug it and confirm my theory. I'll do it tomorrow but just wanted to share with you in case you can also have a look  🙏

htdat committed:
Merge branch 'develop' into fix/1195-pending-refund

htdat committed:
Handle cancelled status

htdat committed:
Use the correct status `canceled` (single letter `l` rather than the double one)

htdat commented:
Thanks @mgascam. I handled that `canceled` status here in this commit 0028acff0a1c07a20f9d78242c884fec030f3315 

Basically, the `canceled` status is only possible when refunding from Stripe Dash. It's not possible from WooPayments client side as we have not supported the `canceled` endpoint for refunds yet. Given these facts, the chance to see the `canceled` status is very low but anyway, we could support it without much effort.

mgascam commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -304,34 +283,33 @@ private function process_webhook_refund_updated( $event_body ) {
+		switch ( $status ) {
+			case Refund_Status::FAILED:
+				$this->order_service->handle_failed_refund( $order, $refund_id, $amount, $currency, $matched_wc_refund );
+				if (
```
I think there's an opportunity to handle more `failure_reason` codes as listed in the Stripe [docs](https://docs.stripe.com/api/refunds/object#refund_object-failure_reason) (i.e. `expired_or_canceled_card`). IMO it will be useful for the merchant. But I think we can tackle it on a follow up issue, since we will probably need to refactor how we are handling the failed refund failure codes to use a holistic approach. What do you think? 
Note: the failure reason `expired_or_canceled_card` is already visible in the timeline view

htdat replied:
This is a good point! I agreed that this deserves a seperate issue.

mgascam replied:
I opened https://github.com/Automattic/woocommerce-payments/issues/10645 to tackle it

htdat replied:
Thanks @mgascam for creating the issue.

mgascam commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1872,43 +1923,54 @@ private function generate_dispute_closed_note( $charge_id, $status, $is_inquiry
 
+		$status_text = $is_pending ?
+			sprintf(
+				'<a href="https://woocommerce.com/document/woopayments/managing-money/#pending-refunds" target="_blank" rel="noopener noreferrer">%1$s</a>',
```
```suggestion
				'<a href="https://woocommerce.com/document/woopayments/managing-money/#refunds" target="_blank" rel="noopener noreferrer">%1$s</a>',
```
I could not find a "pending refunds" section in the current docs so I was wondering if maybe we plan to add this section to the docs. If not, then maybe we could direct them to the current `#refunds` section of the public docs.

htdat replied:
Yes, it's a part of the post-merge section of this PR. I will contact with PayOps to do this.

> Add documents for this new link https://woocommerce.com/document/woopayments/managing-money/#pending-refunds

htdat replied:
Asked here p1743136317673289-slack-C03EE4CPPK5

mgascam reviewed (approved):
Thanks for addressing my comments @htdat. The code changes look good to me. I appreciate the streamlined process as well of the addition of the refund status constant. I left a comment regarding the public docs url to double-check with you. I also saw an opportunity to enhance the refund failure reason in the order notes, but we can tackle it in a follow up issue.

Manual tests worked as intended for me. I tested both the success and failure flows from the Stripe dash, the Order admin and the timeline and did not find any issues. Great work!
